mtr: T5658: Add VRF support for mtr (+ op_mode wrapper)

bbabich opened the pull request:
## Change Summary
* Adds op_mode wrapper for mtr and all the good things that come along with it (like VRF support).
* Re-introduces 'monitor traceroute' as requested by @c-po 

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
* https://vyos.dev/T5658

## Component(s) name
* mtr
* monitor traceroute

## Proposed changes
* Add op_mode wrapper for mtr to enable ease of use + vrf functionality
* Re-introduce 'monitor traceroute' back to op_mode as requested by @c-po 

## How to test
* monitor traceroute 1.1.1.1
* monitor traceroute vrf mgmt 1.1.1.1
* mtr vrf blah 1.1.1.1


## Checklist:
- [X] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [X] I have linked this PR to one or more Phabricator Task(s)
- [X] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [X] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

bbabich committed:
op-mode: T5658: add VRF support for "monitor traceroute"

c-po commented on src/op_mode/mtr.py:
```diff
@@ -0,0 +1,342 @@
+    'report': {
+        'mtr': '{command} --report',
+        'type': 'noarg',
+        'help': 'This option puts mtr into report mode.  When in this mode, mtr will run for the number of cycles specified by the -c option, and then print statistics and exit.'
```
Please remove the double whitespace after `.`

bbabich replied:
Done

c-po commented on src/op_mode/mtr.py:
```diff
@@ -0,0 +1,342 @@
+    'report-wide': {
+        'mtr': '{command} --report-wide',
+        'type': 'noarg',
+        'help': 'This option puts mtr into wide report mode.  When in this mode, mtr will not cut hostnames in the report.'
```
Please remove the double whitespace after `.`

c-po commented on src/op_mode/mtr.py:
```diff
@@ -0,0 +1,342 @@
+    'displaymode': {
+        'mtr': '{command} --displaymode',
+        'type': '<num>',
+        'help': 'Use  this  option  to  select the initial display mode: 0 (default) selects statistics, 1 selects the stripchart without latency information, and 2 selects the stripchart with latency information.'
```
Multiple stray double whitespaces

bbabich replied:
Done

c-po commented on src/op_mode/mtr.py:
```diff
@@ -0,0 +1,342 @@
+    'show-ips': {
+        'mtr': '{command} --show-ips {value}',
+        'type': '<num>',
+        'help': 'Use  this  option to tell mtr to display both the host names and numeric IP numbers.'
```
Multiple stray double whitespaces

please check all help strings

c-po commented:
In the past (VyOS 1.2) we had already a CLI for mtr under `monitor traceroute` which was removed in by 55d2ee80dcddfbaae8600f6389088f1d718965a8.

I have no strong bounds with `monitor traceroute` or `mtr` on the op-mode tree, but we probably should keep the well known `monitor traceroute` CLI tree as this was already there.

Original Implementation https://github.com/vyos/vyatta-op/blob/crux/templates/monitor/traceroute/node.tag/node.def
which was moved to XML in vyos-1x with f16633f0ef2dd074c24387cbd2b9d5541297509c but later got removed in 55d2ee80dcddfbaae8600f6389088f1d718965a8

Happy to re-add this with the nice CLI helper extension for 1.4 and 1.5 after the stray double whitespaces have been fixed and the CLI was changed to `monitor traceroute`

bbabich commented:
What you suggest makes sense to me too @c-po - I believe all you've requested is now complete. 
Seems the doco doesn't even need updating :) 
https://github.com/vyos/vyos-documentation/blob/master/docs/troubleshooting/index.rst#advanced-connectivity-tests

c-po commented on op-mode-definitions/monitor-traceroute.xml.in:
```diff
@@ -0,0 +1,28 @@
+    <properties>
+      <help>Monitor route with mtr</help>
+    </properties>
+    <children>
```
This can't be build, it results in:

```
Opening and ending tag mismatch: children line 7 and interfaceDefinition, line 28, column 23 (monitor-traceroute.xml, line 28)
```

Please also use two spaces as indent.

bbabich replied:
Fixed and XML verified. Hopefully the last fix :)

c-po commented:
```
Possible completions:
  <hostname>            Traceroute route with mtr
  <x.x.x.x>
  <h:h:h:h:h:h:h:h>


cpo@LR1.wue3:~$ monitor traceroute 1.1.1.1

 Invalid option: 1.1.1.1


Possible completions:
  <Enter>               Execute the current command
  <nocomps>             mtr options


cpo@LR1.wue3:~$ monitor traceroute 1.1.1.1
mtr: Unknown host: traceroute
```

Was this tested locally?

c-po committed:
op-mode: T5658: adjust "monitor traceroute" CLI argument index

c-po commented:
`monitor traceroute 1.1.1.1` now works with [709c578 ](https://github.com/vyos/vyos-1x/commit/709c578e123bcf258eba1d15842b63eb63413523) - CLI completion helper is still broken.

c-po committed:
T5658: add common methods interface_list() and vrf_list() to vyos.utils.network

Reduce amount of duplicated (3 times) code in op-mode scripts for ping,
traceroute and mtr.

c-po committed:
op-mode: T5658: reduce amount of exposed options

Example: we should focus on JSON output and not expose XML and CSV.

c-po committed:
op-mode: T5658: fix "monitor traceroute" completion helper

c-po commented:
@MergifyIo backport sagitta

bbabich commented:
> `monitor traceroute 1.1.1.1` now works with [709c578 ](https://github.com/vyos/vyos-1x/commit/709c578e123bcf258eba1d15842b63eb63413523) - CLI completion helper is still broken.
> 
> 

Thanks for finishing this off and apologies for not sorting it out myself. A better outcome with your additional cleanup/fixes too anyhow.

I'll do better next time...
